hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/io/hops/yarn/server/resourcemanager/quota/QuotaService.java

```diff
@@ -0,0 +1,454 @@
+                  container.getContainerState())));
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  public void computeQuota(ContainerId containerId, long computeTime) {
+    if (quotaServiceEnabled) {
+      dispatcher.getEventHandler().handle(
+          new QuotaComputeEvent(containerId, computeTime));
+    }
+  }
+
+  private void addToRunningContainers(QuotaContainerStartEvent event) {
+    synchronized (event.getContainerStartData()) {
+      runningContainers.put(event.getContainerId(), event.getContainerStartData());
+      containerQuotaTime.put(event.getContainerId(), event.getContainerStartData().getStartTime());
+    }
+  }
+
+  private void processContainerFinish(QuotaContainerFinishEvent event) throws IOException {
+    ContainerStartData startData = runningContainers.remove(event.getContainerId());
+    if (startData != null) {
+      computeAndApplyCharge(startData, event.getContainerFinishData().getFinishTime(), true);
+    } else {
+      LOG.error("Container finished before starting: " + event.getContainerId());
+    }
+  }
+
+  private void processContainerUpdate(QuotaContainerUpdateEvent event) throws IOException {
+    ContainerStartData startData = runningContainers.put(event.getContainerId(), event.getContainerStartData());
+    if (startData != null) {
+      computeAndApplyCharge(startData, event.getUpdateTime(), false);
+    } else {
+      LOG.error("Container updated before starting: " + event.getContainerId());
+    }
+  }
+
+  private void processQuotaCompute(QuotaComputeEvent event) throws IOException {
+    ContainerStartData startData = runningContainers.get(event.getContainerId());
+    if (startData != null) {
+      computeAndApplyCharge(startData, event.getComputeTime(), false);
+    } else if (LOG.isDebugEnabled()) {
+      LOG.debug("comput event received after container finished");
+    }
+  }
+
+  private void computeAndApplyCharge(final ContainerStartData startData, long endTime, boolean remove) throws
+      IOException {
+    synchronized (startData) {
+      Long quotaTime = new Long(0);
+      if (remove) {
+        quotaTime = containerQuotaTime.remove(startData.getContainerId());
+        if (endTime - startData.getStartTime() < minRunTime) {
+          endTime = startData.getStartTime() + minRunTime - quotaTime;
+        }
+      } else {
+        quotaTime = containerQuotaTime.get(startData.getContainerId());
+      }
+      if (quotaTime == null) {
+        LOG.error("No container should reach this point without having a quotaTime");
+        return;
+      }
+
+      final Map<String, Long> resourcesMap = new HashMap<>();
+
+      long usedMillis = endTime - quotaTime;
+      Resource resource = startData.getAllocatedResource();
+      for (ResourceInformation entry : resource.getResources()) {
+        resourcesMap.put(entry.getName(), entry.getValue());
+      }
+
+      LightWeightRequestHandler quotaSchedulerHandler = new LightWeightRequestHandler(YARNOperationType.OTHER) {
```

Comment:
Is it safe if we moved DB operation outside of `synchronized` block and let the database handle concurrency?

Reply:
I will remove the synchronized as all call for a same container should be handled by a same thread.